mauth_client/middlewares/asgi.py

```diff
@@ -0,0 +1,86 @@
+import json
+import logging
+
+from asgiref.typing import (
+    ASGI3Application,
+    ASGIReceiveCallable,
+    ASGISendCallable,
+    Scope,
+)
+
+from mauth_client.authenticator import LocalAuthenticator
+from mauth_client.config import Config
+from mauth_client.signable import RequestSignable
+from mauth_client.signed import Signed
+from mauth_client.utils import decode
+
+logger = logging.getLogger("mauth_asgi")
+
+
+class MAuthASGIMiddleware:
+    def __init__(self, app: ASGI3Application) -> None:
+        self._validate_configs()
+        self.app = app
+
+    async def __call__(
+        self, scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable
+    ) -> None:
+        url = scope["path"]
+        query_string = scope["query_string"]
+
+        if query_string:
+            url += f"?{decode(query_string)}"
+
+        headers = {decode(k): decode(v) for k, v in scope["headers"]}
+        body = await self._get_body(receive)
+
+        signable = RequestSignable(
+            method=scope["method"],
+            url=url,
+            body=body,
+        )
+        authenticator = LocalAuthenticator(
+            signable,
+            Signed.from_headers(headers),
+            logger,
+        )
+
+        is_authentic, status, message = authenticator.is_authentic()
+
+        if is_authentic:
+            # asgi spec calls for passing a copy of the scope rather than mutating it
+            # note: deepcopy will blow up with infi recursion due to objects in some values
+            scope_copy = scope.copy()
+            scope_copy["mauth"] = {"app_uuid": authenticator.signed.app_uuid}
```

Comment:
i went with slapping the app uuid into `scope["mauth"]["app_uuid"]` but open to other suggestions.. perhaps it should be top-level?  i think in ruby client we do top-level with key `mauth.app_uuid`.

Reply:
also, looking at the ruby client now, there is other stuff that it sticks into the context also: `mauth.protocol_version` and `mauth.authentic`.  the latter seems kinda useless, but the protocol version might be worthwhile?

Reply:
decided to just go with same context keys we are setting in the ruby middleware